src/object.c

```diff
@@ -135,6 +150,148 @@ robj *createStringObject(const char *ptr, size_t len) {
+    if (oldkey != NULL) sdsfree(oldkey);
+    *(sds *)data = sdsdup(key);
+}
+
+/* Converts (updates, possibly reallocates) 'val' to a valkey object by
+ * attaching a key to it. This functions takes ownership of "one refcount" of
+ * val. Think of val's refcount being decremented by one and the returned
+ * object's refcount being incremented by one. Confused? Simply use the returned
+ * object instead of 'val' after calling this function and you'll be fine. */
+valkey *objectConvertToValkey(robj *val, const sds key) {
+    /* If a key pointer is already embedded, free that sds string first. */
+    if (val->hasembkeyptr) {
+        /* Find the correct location in val's data field. */
+        unsigned char *data = (void *)(val + 1);
+        if (val->hasexpire) {
+            /* Skip expire field */
+            data += sizeof(long long);
+        }
+        sds oldkey = *(sds *)data;
+        sdsfree(oldkey);
+        *(sds *)data = NULL;
+    }
+
+    if (val->encoding == OBJ_ENCODING_EMBSTR) {
+        /* Create a new object with the key embedded and return it. */
+
+        /* TODO: If there's space in val's allocation, we can embed the key
+         * there and memmove the the embedded value, without creating a new
+         * object.
+         *
+         * TODO: If key + value are too large (allocation > 64 bytes) we may not
+         * want to embed both of them. We can embed one or the other depending
+         * on sizes. */
+
+        /* Create a new object with val and key embedded and decrement the
+         * reference counter of 'val'. */
+
+        /* Calculate sizes */
+        size_t key_sds_size = sdscopytobuffer(NULL, 0, key, NULL);
+        size_t val_len = sdslen(val->ptr);
+
+        size_t min_size = sizeof(robj);
+        min_size += sizeof(long long); /* expire */
+        /* Size of embedded key, incl. 1 byte for prefixed sds hdr size. */
+        min_size += 1 + key_sds_size;
+        /* Size of embedded value (EMBSTR) including \0 term. */
+        min_size += sizeof(struct sdshdr8) + val_len + 1;
+
+        size_t bufsize = 0;
+        valkey *o = zmalloc_usable(min_size, &bufsize);
+        o->type = val->type;
+        o->encoding = val->encoding;
+        o->refcount = 1;
+        o->lru = val->lru;
+        o->hasexpire = 1;
+        o->hasembkey = 1;
+        o->hasembkeyptr = 0;
+
+        /* Set the embedded data. */
+        unsigned char *data = (void *)(o + 1);
+
+        /* Set the expire field. */
+        long long expire = -1;
+        *(long long *)data = expire;
+        //memcpy(data, &expire, sizeof(long long));
+        data += sizeof(long long);
+
+        /* Copy embedded string. */
+        sdscopytobuffer(data + 1, key_sds_size, key, data);
+        data += 1 + key_sds_size;
+
+        /* Copy embedded value (EMBSTR). */
+        struct sdshdr8 *sh = (void *)data;
+        sh->flags = SDS_TYPE_8;
+        sh->len = val_len;
+        size_t capacity = bufsize - (min_size - val_len);
```

Comment:
Yes. It's old code, just moved/copied around. You're welcome to suggest simplification, or do it as a follow-up. I wanted to focus on getting things to work first and to avoid changing more than necessary to keep the diff less huge.

Reply:
Do you have a suggestion for a code comment or different variable names?